text/0145-remove-unnecessary-allocator-usage.md

```diff
@@ -0,0 +1,718 @@
+
+Furthermore, the existence of the host-side allocator has become questionable over time. It is implemented in a very naive way, and for determinism and backwards compatibility reasons, it needs to be implemented exactly identically in every client implementation. Runtimes make substantial use of heap memory allocations, and each allocation needs to go through the runtime <-> host boundary twice (once for allocating and once for freeing). Moving the allocator to the runtime side would be a good idea, although it would increase the runtime size. But before the host-side allocator can be deprecated, all the host functions that use it must be updated to avoid using it.
+
+## Stakeholders
+
+No attempt was made to convince stakeholders.
+
+## Explanation
+
+### New definitions
+
+#### <a name="new-def-i"></a>New Definition I: Runtime Optional Positive Integer
+
+The Runtime optional positive integer is a signed 64-bit value. Positive values in the range of [0..2³²) represent corresponding unsigned 32-bit values. The value of `-1` represents a non-existing value (an _absent_ value). All other values are invalid.
+
+#### <a name="new-def-ii"></a>New Definition II: Runtime Optional Pointer-Size
+
+The runtime optional pointer-size has exactly the same definition as runtime pointer-size ([Definition 216](https://spec.polkadot.network/chap-host-api#defn-runtime-pointer-size)) with the value of 2⁶⁴-1 representing a non-existing value (an _absent_ value).
+
+### Changes to host functions
+
+#### ext_storage_get
+
+The function is deprecated. Users are encouraged to use `ext_storage_read_version_2` instead.
+
+#### ext_storage_read
+
+The new version 2 is introduced, deprecating `ext_storage_read_version_1`. The new signature is
+
+```wat
+(func $ext_storage_read_version_2
+    (param $key i64) (param $value_out i64) (param $value_offset i32) (result i64))
+```
+
+##### Arguments
+
+* `key` is a pointer-size ([Definition 216](https://spec.polkadot.network/chap-host-api#defn-runtime-pointer-size)) to the storage key being read;
+* `value_out` is a pointer-size ([Definition 216](https://spec.polkadot.network/chap-host-api#defn-runtime-pointer-size)) to a buffer where the value read should be stored. If the buffer is not long enough to accommodate the value, the value is truncated to the length of the buffer;
+* `value_offset` is a 32-bit offset from which the value reading should start.
+
+##### Result
+
+The result is an optional positive integer ([New Definition I](#new-def-i)), representing either the full length of the value in storage or the _absence_ of such a value in storage.
+
+##### Changes
+
+The logic of the function is unchanged since the previous version. Only the result representation has changed.
```

Comment:
For each of the host function can you please write what exactly has changed vs the previous version and *why*?

And isn't the logic actually changed? (:

Like, in this case:

 - `ext_storage_read_version_1` unnecessarily returns the length as a SCALE-encoded value inside of a host-allocated buffer; version 2 directly returns it as a primitive value.
 - `ext_storage_read_version_1` returns the number of bytes *left* at the supplied offset, and here you've made it to always return the "the full length". (Isn't this a change to the logic?)